File: vistaone-api/app/blueprints/services/well_service.py
from uuid import UUID
from app.models.well import Well
from app.models.user import User
from app.blueprints.repository.well_repository import WellRepository
import logging

logger = logging.getLogger(__name__)


class WellService:
    @staticmethod
    def search_wells(search_text, status, page, per_page, sort_by, order, client_id=None):
        return WellRepository.search(
            search_text=search_text,
            status=status,
            page=page,
            per_page=per_page,
            sort_by=sort_by,
            order=order,
            client_id=client_id,
        )

    @staticmethod
    def create_well(well: Well, current_user_id: str):
        try:
            user = User.query.get(current_user_id)
            if not user or not user.client_id:
                raise ValueError("Authenticated user has no associated client")
            well.client_id = user.client_id
            well.created_by = current_user_id
            well.updated_by = current_user_id
            return WellRepository.create(well)
        except Exception as e:
            logger.error(f"Error creating well: {str(e)}")
            raise e

    @staticmethod
    def get_well(well_id: UUID | str, client_id=None):
        well = WellRepository.get_by_id(well_id, client_id=client_id)
        if not well:
            raise ValueError("Well not found")
        return well

    @staticmethod
    def get_all_wells(client_id=None):
        return WellRepository.get_all(client_id=client_id)

    @staticmethod
    def update_well(well: Well, current_user_id: str):
        try:
            well.updated_by = current_user_id
            return WellRepository.update(well)
        except Exception as e:
            logger.error(f"Error updating well: {str(e)}")
            raise e

    @staticmethod
    def delete_well(well_id: UUID | str, client_id=None):
        try:
            result = WellRepository.delete(well_id, client_id=client_id)
            if not result:
                raise ValueError("Well not found")
            return True
        except Exception as e:
            logger.error(f"Error deleting well: {str(e)}")
            raise e

File: vistaone-web/src/services/wellService.js
import { authFetch } from './apiClient';

const WELL_ENDPOINT = '/wells';

async function parseError(res, fallback) {
  const data = await res.json().catch(() => ({}));
  throw new Error(data?.message || fallback);
}

export async function getWells() {
  const res = await authFetch(WELL_ENDPOINT, { method: 'GET' });
  if (!res.ok) await parseError(res, 'Failed to fetch wells');
  return await res.json();
}

export async function searchWells({ q = '', status = '', page = 1, per_page = 10, sort_by = 'created_at', order = 'desc' } = {}) {
  const params = new URLSearchParams();
  if (q) params.set('q', q);
  if (status) params.set('status', status);
  params.set('page', String(page));
  params.set('per_page', String(per_page));
  params.set('sort_by', sort_by);
  params.set('order', order);
  const res = await authFetch(`${WELL_ENDPOINT}/search?${params.toString()}`, { method: 'GET' });
  if (!res.ok) await parseError(res, 'Failed to search wells');
  return await res.json();
}

export async function createWell(data) {
  const res = await authFetch(WELL_ENDPOINT + '/', {
    method: 'POST',
    body: JSON.stringify(data),
  });
  if (!res.ok) await parseError(res, 'Failed to create well');
  return await res.json();
}

export async function updateWell(id, data) {
  const res = await authFetch(`${WELL_ENDPOINT}/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });
  if (!res.ok) await parseError(res, 'Failed to update well');
  return await res.json();
}

export async function deleteWell(id) {
  const res = await authFetch(`${WELL_ENDPOINT}/${id}`, { method: 'DELETE' });
  if (!res.ok) await parseError(res, 'Failed to delete well');
  return await res.json();
}
